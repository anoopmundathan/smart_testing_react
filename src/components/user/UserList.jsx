/* eslint react/jsx-no-bind: 0 */
/* eslint react/prop-types: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { List, ListItem } from 'material-ui/List'
import ActionGrade from 'material-ui/svg-icons/action/grade'
import Avatar from 'material-ui/Avatar'
import { pinkA200 } from 'material-ui/styles/colors'
import { LARGE } from '../../utils/WithWidth'

const UserList = (props) => {
  const listItems = props.users.map(user => (
    <ListItem
      onClick={props.showUserProfile.bind(this, user)}
      key={user.username}
      style={{ color: 'black' }}
      primaryText={`${user.name.first} ${user.name.last}`}
      leftIcon={<ActionGrade color={pinkA200} />}
      rightAvatar={<Avatar src={`images/${user.username}_sm.jpg`} />}
    />
  ))

  const listView = <List className="view">{listItems}</List>
  const childrenView = <div className="view">{props.children}</div>

  let masterView,
    detailView

  if (props.username && props.width !== LARGE) {
    masterView = childrenView
    detailView = null
  } else {
    masterView = listView
    detailView = childrenView
  }
  return (
    <div className="view-container">
      {masterView}
      {detailView}
    </div>
  )
}

UserList.propTypes = {
  users: PropTypes.array.isRequired,
  width: PropTypes.number.isRequired,
  showUserProfile: PropTypes.func.isRequired
}

export default UserList
