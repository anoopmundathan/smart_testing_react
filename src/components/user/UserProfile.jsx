/* eslint react/prop-types: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { Card, CardActions, CardHeader, CardMedia, CardTitle } from 'material-ui/Card'

const UserProfile = ({ user, username }) => {
  let fullname = ''
  let email = ''
  let gender = ''

  if (user) {
    fullname = `${user.name.title} ${user.name.first} ${user.name.last}`
    email = user.email
    gender = user.gender
  }

  return (
    <Card>
      <CardHeader title={fullname} subtitle={username} avatar={`/images/${username}_sm.jpg`} />
      <CardMedia overlay={<CardTitle title={email} subtitle={gender} />}>
        <img alt="image_username" src={`/images/${username}_lg.jpg`} />
      </CardMedia>
      <CardActions />
    </Card>
  )
}

UserProfile.propTypes = {
  user: PropTypes.object.isRequired
}

export default UserProfile
