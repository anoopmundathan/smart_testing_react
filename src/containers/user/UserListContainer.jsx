/* eslint no-console: 0 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'
import UserList from '../../components/user/UserList'
import * as api from '../../api'
import * as actions from '../../actions/users'
import withWidth from '../../utils/WithWidth'

export class UserListContainer extends Component {
  componentDidMount () {
    // Imagine we want to fetch users from another component. How do you think
    // we could reuse this code?
    api
      .getUsers()
      .then((users) => {
        this.props.dispatch(actions.receiveUsers(users))
      })
      .catch((err) => {
        console.log(err)
      })
  }

  showUserProfile = (user) => {
    this.props.router.push(`/users/${user.username}`)
  };

  render () {
    return (
      <UserList
        {...this.props}
        username={this.props.params.username}
        users={this.props.users}
        showUserProfile={this.showUserProfile}
        width={this.props.width}
      />
    )
  }
}

UserListContainer.propTypes = {
  params: PropTypes.object.isRequired,
  width: PropTypes.number.isRequired,
  router: PropTypes.object,
  users: PropTypes.array,
  dispatch: PropTypes.func
}

UserListContainer.defaultProps = {
  users: []
}

const mapStateToProps = state => ({
  users: state.users
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = dispatch => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(mapStateToProps, mapDispatchToProps)(
  withRouter(withWidth(UserListContainer))
)
