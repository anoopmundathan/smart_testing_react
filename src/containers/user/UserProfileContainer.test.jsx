import React from 'react'
import getMuiTheme from 'material-ui/styles/getMuiTheme'
import FakeConfigureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import ConnectedUserProfileContainer, { UserProfileContainer } from './UserProfileContainer'
import { mockFetchContainer } from '../../testUtils'
import ActualConfigureStore from '../../store'
import * as actions from '../../actions/user'

describe('<UserProfileContainer />', () => {
  it('should render with default props', () => {
    const props = {
      params: { username: '123' }
    }

    const wrapper = shallow(<UserProfileContainer {...props} />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  const userData = {
    user: {
      name: {
        title: 'mr',
        first: 'Eduard',
        last: 'lama'
      },
      email: 'eduard@gmail.com',
      gender: 'male'
    }
  }

  const initialState = {
    user: {
      name: {
        title: 'mr',
        first: 'Manolo',
        last: 'lama'
      },
      email: 'manolo@gmail.com',
      gender: 'male'
    }
  }
  const mockStore = FakeConfigureStore()
  const props = {
    params: { username: '123' }
  }
  let store

  describe('<ConnectedUserProfileContainer />', () => {
    it('should render connected component', () => {
      store = mockStore(initialState)
      const container = shallow(<ConnectedUserProfileContainer {...props} store={store} />)

      expect(shallowToJson(container)).toMatchSnapshot()
    })
  })

  describe('Integration test: Mount + Wrapping Provider', () => {
    let wrapper

    beforeEach(() => {
      store = mockStore(initialState)

      mockFetchContainer(userData.user)

      wrapper = mount(
        <Provider store={store}>
          <ConnectedUserProfileContainer {...props} />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes: {
            muiTheme: React.PropTypes.object.isRequired
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedUserProfileContainer).length).toEqual(1)
    })

    it('should Prop matches with initialState', () => {
      expect(wrapper.find(UserProfileContainer).prop('user')).toEqual(initialState.user)
    })

    it('should check action dispatch', () => {
      store.dispatch(actions.receiveUser(userData.user))
      const action = store.getActions()

      expect(action[0].type).toBe(actions.RECEIVE_USER)
    })
  })

  describe('Integration test: Mount + Wrapping Provider + Actual Store + Reducers', () => {
    let wrapper

    beforeEach(() => {
      store = ActualConfigureStore(initialState)

      mockFetchContainer(userData.user)

      wrapper = mount(
        <Provider store={store}>
          <ConnectedUserProfileContainer {...props} />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes: {
            muiTheme: React.PropTypes.object.isRequired
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedUserProfileContainer).length).toEqual(1)
    })

    it('should check action dispatch', () => {
      // OR store.dispatch(actions.receiveUser(userData))
      wrapper.find(UserProfileContainer).props().dispatch(actions.receiveUser(userData.user))

      expect(wrapper.find(UserProfileContainer).prop('user')).toEqual(userData.user)
    })
  })
})
