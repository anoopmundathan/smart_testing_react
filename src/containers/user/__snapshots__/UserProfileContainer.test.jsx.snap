// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfileContainer /> <ConnectedUserProfileContainer /> should render connected component 1`] = `
<UserProfileContainer
  dispatch={[Function]}
  params={
    Object {
      "username": "123",
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  user={
    Object {
      "email": "manolo@gmail.com",
      "gender": "male",
      "name": Object {
        "first": "Manolo",
        "last": "lama",
        "title": "mr",
      },
    }
  }
/>
`;

exports[`<UserProfileContainer /> should render with default props 1`] = `
<UserProfile
  params={
    Object {
      "username": "123",
    }
  }
  user={Object {}}
  username="123"
/>
`;
